### ISSUE TITLE
Better support for enum-like classes

### ISSUE THREAD

**Issue by otaviomacedo** [cdklabs/decdk#201]
Enum-like classes can be ambiguous. For example, the `lambda.Function` class has a property `code: Code`. When deCDK finds a template like:

```json
"HelloLambda": {
  "Type": "aws-cdk-lib.aws_lambda.Function",
  "Properties": {
    "code": {
      "fromAsset": { "path": "." }
    },
  },
  ...
}
```

it currently interprets `"fromAsset"` as `Code.fromAsset()`, which is fine. But if a new enum-like class is introduced to the construct library (let's say, `SpecialCode`), also with a `SpecialCode.fromAsset()` method, there is no way to specify which enum-like we want.

Also, if the type of the property is another class or an interface (e.g., `code: ICode`), deCDK can't find the enum-like class that has the desired static method.

The obvious solution is to require the fully qualified name of the static method (`"aws_cdk_lib.aws-lambda.Code.fromAsset"` in this case). But this tends to add a lot of noise to the templates. We need to find the right UX for this feature.

**Comment by mrgrain** [comment]
> Also, if the type of the property is another class or an interface (e.g., code: ICode), deCDK can't find the enum-like class that has the desired static method.

A good example for this is [machineImage](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_autoscaling.AutoScalingGroup.html#machineimage) on the AutoScalingGroup.

It requires `IMachineImage` which doesn't provide any implementation on how to get a `MachineImage`. Instead we have an Enum-like helper `MachineImage`

```
  "ASG": {
    "Type": "aws-cdk-lib.aws_autoscaling.AutoScalingGroup",
    "Properties": {
      "vpc": { "Ref": "VPC" },
      "machineImage": {
        "MachineImage.latestAmazonLinux": {}
      }
    }
  },
```

as well as independent classes:
```
      "machineImage": {
        "AmazonLinuxImage": {}
      }
```

**Comment by mrgrain** [comment]
> The obvious solution is to require the fully qualified name of the static method ("aws_cdk_lib.aws-lambda.Code.fromAsset" in this case). But this tends to add a lot of noise to the templates. We need to find the right UX for this feature.

I could imagine this as searching upwards in the module path, starting from the location of the current Resource.

E.g in the lambda example above, the minimal thin to write would be:
```
"Type": "aws-cdk-lib.aws_lambda.Function",
"Properties": {
  "code": {
    "Code.fromAsset": { "path": "." }
  }
}
```
which would mean we are looking for an enum/class in the following order:
- `aws-cdk-lib.aws_lambda.Code` <--- this would be found
- `aws-cdk-lib.Code`
- `Code` (this is basically checking if it is a FQN)

We could still allow omitting the first `Code.` before `fromAsset`. In that case the first lookup attempt would be something like: `(typeof aws-cdk-lib.aws_lambda.FunctionProps).fromAsset`

This would apply to many common scenarios, allowing users to simply use the enum-like class name (or even omitting it).

---

In a more complex scenario, when the enum-like is cross module:
```
"Type": "aws-cdk-lib.aws_autoscaling.AutoScalingGroup",
"Properties": {
  "machineImage": {
    "aws_ec2.MachineImage.latestAmazonLinux": {}
  }
}
```
lookup would be in the following order:
- `aws-cdk-lib.aws_autoscaling.aws_ec2.MachineImage`
- `aws-cdk-lib.aws_ec2.MachineImage` <--- this would be found
- `aws_ec2.MachineImage`

---
Bonus: When we can use a non-construct class to represent the interface. E.g. `AmazonLinuxImage` in the example above or `AccountPrincipal`.

```
"Type": "aws-cdk-lib.aws_autoscaling.AutoScalingGroup",
"Properties": {
  "machineImage": {
    "aws_ec2.AmazonLinuxImage": {}
  }
}
```
Same lookup order, but somehow we cannot drop the last part:
- `aws-cdk-lib.aws_autoscaling.aws_ec2.AmazonLinuxImage`
- `aws-cdk-lib.aws_ec2.AmazonLinuxImage` <--- this would be found
- `aws_ec2.AmazonLinuxImage`

We then need to check if the class implements the required interface.

**Comment by mrgrain** [comment]
Fixed in #241 and #242
